addons/website/controllers/form.py

```diff
@@ -303,8 +303,10 @@ def insert_record(self, request, model, values, custom, meta=None):
             # If there is a default field configured for this model, use it.
             # If there isn't, put the custom data in a message instead
             if default_field.name:
-                if default_field.ttype == 'html' or model_name == 'mail.mail':
```

Comment:
no, `html_sanitize` won't use `nl2br_enclose` nor `nl2br`

sementically, when you call `nl2br_enclose` explictly here without any comment to claim it as a hack, you expect the value flushed to the DB should be for one single html element since it is enclosed.

But in the test `TestWebsiteCrm.test_tour`,  the value it expects and searches is `'<p>EXTERNAL SUBMISSION - Customer not verified<br>\n<br>\n</p><p>### TOUR DATA ###</p>'` which is for 2 elements.

this is because the written value
`'<div>EXTERNAL SUBMISSION - Customer not verified<br>\n<br>\n<p>### TOUR DATA ###</p></div>'`
is sanitized once to cache as
`'EXTERNAL SUBMISSION - Customer not verified<br>\n<br>\n<p>### TOUR DATA ###</p>'`
and sanitized twice to db as
`'<p>EXTERNAL SUBMISSION - Customer not verified<br>\n<br>\n</p><p>### TOUR DATA ###</p>'`

my question is: what do we expect the value stored in db, shoud it be one single element?
if so, we should change the test
if not, we should use `nl2br` like this commit

Also could you help me check if the `nl2br_enclose` was added on purpose for `model_name == 'mail.mail'`?

Reply:
> my question is: what do we expect the value stored in db, shoud it be one single element?

I don't think we should ever expect that no. And yes, I don't really understand in which case `nl2br_enclose` would be needed...

> Also could you help me check if the nl2br_enclose was added on purpose for model_name == 'mail.mail'?

... although it seems to have been added on purpose with https://github.com/odoo/odoo/commit/c49459f70df59764f6ee5d12fb59e6c09e1c21fd, did you check that? Maybe @bso-odoo remembers and can help otherwise. The whole diff is about using `nl2br_enclose`  to encode the saved HTML, probably that should just have been encoded with the wrapping div induced by `nl2br_enclose` ? This was also made as a fix in the oldest stable version, it might need to be re-changed there as well :fearful:

Reply:
Mini summary of the discussion:
at the time that security fix was made, `nl2br` did not escape the HTML, but `nl2br_enclose` did - so this is the method that got used, it accidentally introduced the side-effect of wrapping the HTML into an extra `<div>`, but that was not actually wanted.